### DISCUSSION TITLE
Add possibility to provide schema using Promises

### DISCUSSION THREAD

**Discussion by ShockOfficial** [next-safe-action/next-safe-action#151] [Ideas]
Hi!
Something that really bothers me is that there is no way to provide schema asynchronously. 
In the project, I am using `next-intl` with `zod` to give error messages in the appropriate language so basically the way to do that is to provide the translation function while creating the schema
```ts
export const createAddSpaceSchema = (
	t: TranslationFunction<'AddSpaceModalValidation'>,
) =>
	z.object({
		name: z
			.string()
			.min(3, { message: t('name_min') })
			.max(64, { message: t('name_max') }),
		description: z.string().min(1, { message: t('description_required') }),
		type: z.enum(['public', 'private']),
		tags: z.array(z.object({ value: z.string() })),
	});
```
On the server side, there is async `getTranslations()` function that can be used to get translations but the `schema()` can't handle Promises.  
```ts
export const createSpace = action
	.schema(async () => {
		const t = await getTranslations('AddSpaceModalValidation');
		return createAddSpaceSchema(t);
	})
	.action(async (data) => {
		console.log('data', data);
		return {
			id: 1,
			name: 'test',
		};
	});
```

Something that I would appreciate is to have an interface for schema that accepts and handles Promise or other way to provide schema later on using maybe middleware with `use` and passing schema through `ctx`  that would be typesafe 

My current approach works but is not ideal since this is not type-safe on the client side. 

```ts
export const createSpace = action
	.use(async ({ next, ...rest }) => {
		const t = await getTranslations('AddSpaceModalValidation');
		const schema = createAddSpaceSchema(t);
		console.log('rest.clientInput', rest.clientInput);
		const parsedData = schema.parse(rest.clientInput);
		return next({ ...rest, ctx: { schema, parsedData } });
	})
	.action(async (data) => {
		const { parsedData } = data.ctx;

		console.log('parsedData', parsedData);

		return {
			id: 1,
			...parsedData,
		};
	});
```

If there is no quick fix let me know. Will try to refactor the way I create schema with translations.

**Comment by TheEdoRan** [comment]
Check out #111, thanks.
